Parser/Tree_Nodes/r_asmt.c

```diff
@@ -24,27 +24,45 @@ r_asmt *create_r_asmt(GArray *tokenStream, unsigned long index, unsigned long *n
 
     r_asmt *new_asmt = calloc(1, sizeof(r_asmt));
 
-    // Need to try expression before we add the ID, because it cannot be self-referencing
-    new_asmt->expression = create_expr(tokenStream, index + 2, next);
-
-    Type *varType = malloc(sizeof(Type));
-
+    //Get expression so we know the type
     // Get the variable in the reassignment
-    Token *curToken = &g_array_index(tokenStream, Token, index);
-
+    Token *curToken = &g_array_index(tokenStream, Token, (*next));
+    // Hold the type of the ID, will compare later
+    Type t;
     // Does the ID exist, because we cant reassign a non existant assignment
-    if (!findIDType(curToken->data, varType)) {
-        // We tried to redfine a variable that doesnt exit
-        fprintf(stderr, "Syntax Error: Tried to reassign a variable that does not exist");
+    if (!findIDType(curToken->data, &t)) {
+        // We tried to redefine a variable that doesnt exit
+        fprintf(stderr, "Syntax Error: Incorrect reassignment");
         exit(-1);
     }
-
     new_asmt->id = create_id_node(curToken);
-    curToken = &g_array_index(tokenStream, Token, index + 1);
+    (*next)++;
+    curToken = &g_array_index(tokenStream, Token, (*next));
     if (curToken->type != t_assign) {
         fprintf(stderr, "Syntax Error: Unexpected Token in reassignment");
         exit(-1);
     }
+    (*next)++;
+    new_asmt->expression = create_expr(tokenStream, (*next), next);
+
+    if (t == jdouble || t == jf_double) {
+        if (new_asmt->expression->double_expression == NULL) {
+            fprintf(stderr, "Syntax Error: Incorrect reassignment, types do not match.");
+            exit(-1);
+        }
+    }
+    if (t == jint || t == jf_int) {
+        if (new_asmt->expression->int_expression == NULL) {
+            fprintf(stderr, "Syntax Error: Incorrect reassignment, types do not match.");
+            exit(-1);
+        }
+    }
+    if (t == jstring || t == jf_str) {
+        if (new_asmt->expression->string_expression == NULL) {
+            fprintf(stderr, "Syntax Error: Incorrect reassignment, types do not match.");
+            exit(-1);
+        }
+    }
```

Comment:
This looks good, only thing is I would verify that the re-assignment ends with a semicolon, this is where you would use the `next` pointer.